### ISSUE TITLE
protoc-gen-openapiv2: required fields not optional in partial update operations

### ISSUE THREAD

**Issue by oyvindwe** [grpc-ecosystem/grpc-gateway#2566]
## 🐛 Bug Report

Currently a field annotated with `(google.api.field_behavior) = REQUIRED` is marked as required in all request messages.
However, according to [Goole AIP-203](https://google.aip.dev/203#required) they may not be required for partial updates using `PATCH` and `update_mask`:

> When [updating](https://google.aip.dev/134) the resource, the user **may** omit the field provided that the field is also absent from the field mask, indicating no change to the field (otherwise it **must** be provided).

## To Reproduce

1. Create `example/foo.proto` with:
```
syntax = "proto3";

package example;

option go_package = "example.com/foo";

import "google/api/annotations.proto";
import "google/api/field_behavior.proto";
import "google/protobuf/field_mask.proto";

service FooService {
  rpc CreateFoo(CreateFooRequest) returns (Foo) {
    option (google.api.http) = {
      post: "/foos"
      body: "foo"
    };
  }
  rpc UpdateDataSet(UpdateFooRequest) returns (Foo) {
    option (google.api.http) = {
      patch: "/v1/{foo.name}"
      body: "foo"
    };
  }
}

message CreateFooRequest {
  Foo foo = 1 [(google.api.field_behavior) = REQUIRED];
}

message UpdateFooRequest {
  Foo foo = 1 [(google.api.field_behavior) = REQUIRED];
  google.protobuf.FieldMask update_mask = 2;
}

message Foo {
  string name = 1 [(google.api.field_behavior) = REQUIRED];
  string required_field = 2 [(google.api.field_behavior) = REQUIRED];
  Nested nested  = 3;
}

message Nested {
  string required_field = 1 [(google.api.field_behavior) = REQUIRED];
  string optional_field = 2;
}
```

2. Run `protoc -I . --go_out=. --openapiv2_out . example/foo.proto` (make sure to have the imported packages in the import path)

## Expected behavior

A `foo.swagger.json` file where `required_field` for `Foo` and `Nested` are not required for the `UpdateDataSet` operation.

## Actual Behavior

A `foo.swagger.json` file where `required_field`  for `Foo` and `Nested` are required for the `UpdateDataSet` operation.

## Your Environment

protoc-gen-openapi v2.8.0

**Comment by oyvindwe** [comment]
We're currently working around this problem by not using `[(google.api.field_behavior) = REQUIRED]`, and instead include a comment for fields that are required in create operations.

I think it will be quite complex to implement this (similar to removing path parameters from `body`), as create and update operations requires different OpenAPI types _if_ any field in the data structure is marked as required. We'd also need a rule on how to decide what operations to be considered as partial updates. Any request with a `FieldMask update_mask` on top level? Only `PATCH` bindings?

A possible simpler workaround is to introduce a CLI argument on wheter to respect `[(google.api.field_behavior) = REQUIRED]` at all or not.

**Comment by johanbrandhorst** [comment]
The option to disable `field_behavior = REQUIRED` seems like it wouldn't add much unless the user has added that option for other reasons than the openapiv2 generator, do we have reason to believe that this is the case?

**Comment by oyvindwe** [comment]
I get your point. This issue can probably be viewed as an edge case. In plain gRPC, it's a semantic message to any developer, informing that any `REQUIRED` field must be present when creating a resource or performing a full update on a resource. The issue with mapping this to `required` in OpenAPI, is that the field also becomes required when doing a partial update (e.g. in SwaggerUI, Postman, etc.)

To be fair, Google AIP-203 relaxes the semantics of `REQUIRED` compared with the [annotation documentation](
https://github.com/googleapis/googleapis/blob/aa136fa6649f7d999bf3873d0e5fca75cb2354a5/google/api/field_behavior.proto#L58-L60)

Also, it's not very advantageous to generate separate types for partial updates.

We worked around this issue by removing `(google.api.field_behavior) = REQUIRED` and just stated in the field documentation that it is required for create.

Closing issue.
